src/adaptors/UBImportSMART.cpp

```diff
@@ -0,0 +1,346 @@
+        }
+        tmpNumber++;
+        if (tmpNumber == 100000) {
+            qWarning() << "Import failed. Failed to create temporary directory for iwb file";
+            return "";
+        }
+    }
+    if (!rootDir.mkdir(documentRootFolder)) {
+        qWarning() << "Import failed. Couse: failed to create temp folder for cff package";
+    }
+
+    QFile out;
+    char c;
+    for(bool more=zip.goToFirstFile(); more; more=zip.goToNextFile()) {
+        if(!zip.getCurrentFileInfo(&info)) {
+            //TOD UB 4.3 O display error to user or use crash reporter
+            qWarning() << "Import failed. Cause: getCurrentFileInfo(): " << zip.getZipError();
+            return "";
+        }
+//        if(!file.open(QIODevice::ReadOnly)) {
+//            qWarning() << "Import failed. Cause: file.open(): " << zip.getZipError();
+//            return "";
+//        }
+        file.open(QIODevice::ReadOnly);
+        if(file.getZipError()!= UNZ_OK) {
+            qWarning() << "Import failed. Cause: file.getFileName(): " << zip.getZipError();
+            return "";
+        }
+
+        QString newFileName = documentRootFolder + "/" + file.getActualFileName();
+
+        QFileInfo newFileInfo(newFileName);
+        rootDir.mkpath(newFileInfo.absolutePath());
+
+        out.setFileName(newFileName);
+        out.open(QIODevice::WriteOnly);
+
+        while(file.getChar(&c))
+            out.putChar(c);
+
+        out.close();
+
+        if(file.getZipError()!=UNZ_OK) {
+            qWarning() << "Import failed. Cause: " << zip.getZipError();
+            return "";
+        }
+        if(!file.atEnd()) {
+            qWarning() << "Import failed. Cause: read all but not EOF";
+            return "";
+        }
+
+        file.close();
+
+        if(file.getZipError()!=UNZ_OK) {
+            qWarning() << "Import failed. Cause: file.close(): " <<  file.getZipError();
+            return "";
+        }
+    }
+
+    zip.close();
+
+    if(zip.getZipError()!=UNZ_OK) {
+        qWarning() << "Import failed. Cause: zip.close(): " << zip.getZipError();
+        return "";
+    }
+
+    return documentRootFolder;
+}
+
+
+void UBImportSMART::importSinglePage(UBDocumentProxy* document,
+                                     QByteArray xml)
+{
+    int pageIndex = document->pageCount();
+
+    UBGraphicsScene* scene = UBSvgSubsetAdaptor::loadScene(document, xml);
+    UBApplication::showMessage(tr("Inserting page %1").arg(pageIndex), true);
+
+    // Adjust coordinates of scene items so (0,0) is the
+    // very centre of the page.
+    qreal adjust_x = scene->width() / 2;
+    qreal adjust_y = scene->height() / 2;
+
+    // Do not use the original height of the page, as the
+    // page is scaled to appear on the screen.  Use a 4:3 ratio
+    // instead so the page is large enough to read.
+    qreal newHeight = scene->width() * 0.75;
+    adjust_y -= 0.5 * (scene->height() - newHeight);
+
+    for (QGraphicsItem* item : scene->items())
```

Comment:
Iterating on a non-const container can cause detaches,
```suggestion
    const auto sceneItems = scene->items();
    for (QGraphicsItem* item : sceneItems)
```